vsintegration/src/FSharp.Editor/Options/SettingsPersistence.fs

```diff
@@ -18,42 +18,67 @@ type SVsSettingsPersistenceManager = class end
 type SettingsStore(serviceProvider: IServiceProvider) =
 
     let settingsManager = serviceProvider.GetService(typeof<SVsSettingsPersistenceManager>) :?> ISettingsManager
-    // settings quallified type names are used as keys, this should be enough to avoid collisions
-    let storageKey (typ: Type) = typ.Namespace + "." + typ.Name
+
+    let storageKeyVersions (typ: Type) =
+        // "TextEditor" prefix seems to be required for settings changes to be synced between IDE instances
+        [ "TextEditor.FSharp." + typ.Namespace + "." + typ.Name
+          // we keep this old storage key to upgrade without reverting user changes
+          typ.Namespace + "." + typ.Name ]
+        
+    let storageKey (typ: Type) = storageKeyVersions typ |> List.head
 
     // Each group of settings is a value of some named type, for example 'IntelliSenseOptions', 'QuickInfoOptions'
-    // We cache exactly one instance of each, treating them as immutable.
-    // This cache is updated by the SettingsStore when the user changes an option.
-    let cache = System.Collections.Concurrent.ConcurrentDictionary<Type, obj>()
+    // and it is usually representing one separate option page in the UI.
+    // We cache exactly one immutable value of each type.
+    // This cache is updated by the SettingsStore when the user makes changes in the Options dialog
+    // or when a change is propagated from another VS IDE instance by SVsSettingsPersistenceManager. 
+    let cache = ConcurrentDictionary<Type, obj>()
 
-    let read() =
+    let getCached() =
         match cache.TryGetValue(typeof<'t>) with
-        | true, value -> value :?> 't
-        | _ -> failwithf "Settings %s are not registered." typeof<'t>.Name 
+        | true, (:? 't as value) -> value
+        | _ -> failwithf "Settings %s are not registered." typeof<'t>.Name
 
-    let write settings = cache.[settings.GetType()] <- settings
+    let keepInCache settings = cache.[settings.GetType()] <- settings
+
+    // The settings record, even though immutable, is being effectively mutated in two instances:
+    //   when it is passed to the UI (provided it is marked with CLIMutable attribute);
+    //   when it is being populated from JSON using JsonConvert.PopulateObject;
+    // We make a deep copy in these instances to isolate and contain the mutation
+    let clone (v: 't) = JsonConvert.SerializeObject v |> JsonConvert.DeserializeObject<'t>
 
     let updateFromStore settings =
-        let result, json = settings.GetType() |> storageKey |> settingsManager.TryGetValue
-        if result = GetValueResult.Success then
-            // if it fails we just return what we got
-            try JsonConvert.PopulateObject(json, settings) with _ -> () 
-        settings
-        
-    member __.Read() = read()
+        // make a deep copy so that PopulateObject does not alter the original
+        let copy = clone settings
```

Comment:
I feel like we should not have to do any deep cloning here. We should not be modifying existing immutable objects.

Reply:
If we can find a way to stop modifying immutable objects, we should never have to call `JsonConvert.PopulateObject` and get rid of the `[<CLIMutable>]` attribute on the options. I strongly suggest trying design something with that in mind.

Reply:
Yes, I absolutely agree both of these smell bad. In my defence, at the time I was trying to get something up and running quickly, that would be easy to use so we could have some way to configure things, hoping that in the long run we could tap into the Roslyn infrastructure, but this is still internal, I think.

The proper thing to do here would bo to declare, store and retrieve each and every configurable value separately, the way Roslyn does, that would let us get rid of the PopulateObject stuff. Technically the CLIMutable is not required when someone is willing to write the boilerplate to bind everything to the UI.